### ISSUE TITLE
Emails from the same thread don't generate one threadId via "Email/import".

### ISSUE THREAD

**Issue by zvasilev** [stalwartlabs/jmap-server#25]
Emails from the same thread don't generate one threadId via "Email/import".

Attached are simple sample messages.
Taken from a real thread where the subject was not correct (corrected in the reply).
But in-reply-to header explicitly said it is from this thread.

For protocols which don't provide threadId , an app should do threadId generation.

[msg5_0.txt](https://github.com/stalwartlabs/jmap-server/files/10318034/msg5_0.txt)
[msg5_1.txt](https://github.com/stalwartlabs/jmap-server/files/10318035/msg5_1.txt)

**Comment by mdecimus** [comment]
The problem is that the subjects are different, that is probably why these messages are assigned to different threads.

ms5_0.txt has a subject which already includes the prefix "Subject: ":

``Subject: Subject: message 5``

while msg5_1.txt's subject does not include that prefix:

``Subject: Re: message 5``.

For these messages to be assigned to the same threadId, msg5_1.txt should have the following subject:

``Subject: Re: Subject: message 5``.

**Comment by zvasilev** [comment]
Yep, the subject is not correct. This is just a suggestion.

My observation about threading is that in-reply-to, references and message-id headers, or thread-index for MS Exchange are the only header to build a thread-id.This approach has a caveat though.Some emails in the middle of the thread could arrive later because of server maintenance and anti-spam techniques. So two thread ids could be generated, and when an email in the middle came, the tread ids should be merged.

Subject should be only a fallback mechanism if the previous are missing.
The subject could be something like "RE: Re: Ре: some subject" in different languages. 
So it could be normalized to the last ":" character when building a thread-id.By building thread-id by subject only it is better to have a date limitation. This will prevent false positives to merge new emails to past threads.

**Comment by mdecimus** [comment]
The JMAP server uses this algorithm from RFC8021 to assign threadIds:

>    [...] it is
>    suggested that two messages belong in the same Thread if both of the
>    following conditions apply:
> 
>    1.  An identical message id [[RFC5322](https://datatracker.ietf.org/doc/html/rfc5322)] appears in both messages in any
>        of the Message-Id, In-Reply-To, and References header fields.
>    2.  After stripping automatically added prefixes such as "Fwd:",
>        "Re:", "[List-Tag]", etc., and ignoring white space, the subjects
>        are the same.  This avoids the situation where a person replies
>        to an old message as a convenient way of finding the right
>        recipient to send to but changes the subject and starts a new
>        conversation.

In your example, the messages were assigned to different threads because the word "Subject" is not recognized as a valid prefix that would be added by a mailer. These are the prefixes used by Stalwart JMAP that are commonly added by most clients:

```
static RE_PREFIXES: &[&str] = &[
    "re", "res", "sv", "antw", "ref", "aw", "απ", "השב", "vá", "r", "rif", "bls", "odp", "ynt",
    "atb", "رد", "回复", "转发",
];

static FWD_PREFIXES: &[&str] = &[
    "fwd",
    "fw",
    "rv",
    "enc",
    "vs",
    "doorst",
    "vl",
    "tr",
    "wg",
    "πρθ",
    "הועבר",
    "továbbítás",
    "i",
    "fs",
    "trs",
    "vb",
    "pd",
    "i̇lt",
    "yml",
    "إعادة توجيه",
    "回覆",
    "轉寄",
];

```
The algorithm for extracting the thread name is available at https://github.com/stalwartlabs/mail-parser/blob/main/src/parsers/fields/thread.rs.

**Comment by zvasilev** [comment]
Just trying to run my tests against FM and Stalwart server, and this issue appeared.
It is a simplified test case from a real email thread.
Only in my native language Bulgarian I saw several prefixes like "Ре:", "Отговор:", "Отг.:".
So I decided to not go with hard-coded prefixes.

As I said, it is just a suggestion.
Trying to help :)

**Comment by mdecimus** [comment]
Thank you very much for reporting all these issues you're finding, I really appreciate it!

Regarding the prefixes, I built the prefix table using this list from Wikipedia:

https://en.wikipedia.org/wiki/List_of_email_subject_abbreviations

But probably many others are missing. It's a bit tricky how to implement this, I'll see if I can find a more complete list.

**Comment by zvasilev** [comment]
I have the same bug in my thread builder fallback algorithm.
This is only for IMAP servers that do not provide thread-id.

As a client this is not a priority for me, because I prefer to rely on the servers :)

I thought something like this.
Search backward for ":".
If not found, the entire string is a subject.
Оtherwise if the string to the beginning or next ":" is only one word or 1-2 chars but more than two bytes per char it probably is a prefix.

It could not be a prefix like "Test: a subject". The normalized(stripped) subject will be "a subject"

For clients this will be a problem to not show the entire text , but for thread building it should be enough.
For example a reply will be "Re: Test: a subject". The normalized subject again will be "a subject"

**Comment by mdecimus** [comment]
Hi @zvasilev , happy new year!

Sounds good, I'll review the algorithm once I finish with the other modules I am currently working on.

BTW, please let me know once your JMAP client is ready as I'd like to check it out. Currently I am using Ltt.rs but it has a few bugs that make it hard to use.

**Comment by zvasilev** [comment]
Happy new year too!

I guess it won't be soon. But hope will be less than 1-2 months.

A single test against FM took 30 secs. On my machine with a stalwart server is 3 sec.
So this gave me a lot of speed.
I'm in the middle of refactoring error handling for JMAP.

My data structures are designed around IMAP.

But JMAP "notUpdated", "notDestroyed", etc .. a lot simpler.
No IMAP responses like "OK cmd-id" and in the description "Server is busy, try later".
This is from MS IMAP when try to authenticate :).

**Comment by zvasilev** [comment]
Btw right now the happy path works very well.
I can sync, see, compose emails.
Also archive -> which include -> create an existent folder with JMAP role and move an item into that folder.
Just not handled all error cases properly.
The app intent is not to be just a viewer , so it need to recover from timeouts, ets.

I Also didn't set up "Postfix" so I'm not sure how to send work.
If you wish I would send an invitation to the beta builds?

**Comment by mdecimus** [comment]
> If you wish I would send an invitation to the beta builds?

That would be nice, thanks!
